File: apps/server/api/src/services/git.service.ts
import path from "node:path";
import { execShell as execAsync } from "@/api/src/utils/exec.js";
import { isNotGitRepositoryError } from "@/agent/src/utils/error.js";
import type {
  GitBranchInfo,
  GitBranchesResponse,
  GitFileStatus,
  GitStatusSummary,
} from "@console/types";

export class GitService {
  private parseNumstat(output: string, map: Map<string, { additions: number; deletions: number }>): void {
    for (const line of output.split("\n")) {
      const trimmed = line.trim();
      if (!trimmed) continue;
      const parts = trimmed.split("\t");
      if (parts.length >= 3) {
        const adds = parts[0] === "-" ? 0 : Number.parseInt(parts[0], 10) || 0;
        const dels = parts[1] === "-" ? 0 : Number.parseInt(parts[1], 10) || 0;
        const file = parts[2];
        const existing = map.get(file) || { additions: 0, deletions: 0 };
        map.set(file, {
          additions: existing.additions + adds,
          deletions: existing.deletions + dels,
        });
      }
    }
  }

  private async getNumstatMap(repoPath: string): Promise<Map<string, { additions: number; deletions: number }>> {
    const map = new Map<string, { additions: number; deletions: number }>();
    try {
      const unstaged = await execAsync("git diff --numstat", { cwd: repoPath });
      this.parseNumstat(unstaged.stdout, map);
    } catch {
      // Ignored
    }
    try {
      const staged = await execAsync("git diff --cached --numstat", { cwd: repoPath });
      this.parseNumstat(staged.stdout, map);
    } catch {
      // Ignored
    }
    return map;
  }

  /**
   * Run git status --porcelain=v1 in the repository directory and return structured status.
   */
  async getGitStatus(repoPath: string): Promise<GitStatusSummary> {
    try {
      // Get current branch name
      const branchRes = await execAsync("git rev-parse --abbrev-ref HEAD", { cwd: repoPath });
      const branch = branchRes.stdout.trim() || "main";

      // Get porcelain git status
      const statusRes = await execAsync("git status --porcelain=v1 -u", { cwd: repoPath });
      const lines = statusRes.stdout.split("\n").filter((line) => line.trim().length > 0);

      const numstatMap = await this.getNumstatMap(repoPath);
      const files: import("@console/types").GitFileEntry[] = [];

      for (const line of lines) {
        const indexStatus = line[0];
        const workTreeStatus = line[1];
        const rawFilePath = line.slice(3).trim();

        // Handle renamed files "fileA -> fileB"
        const filePath = rawFilePath.includes("->")
          ? rawFilePath.split("->")[1]!.trim()
          : rawFilePath;

        const absolutePath = path.resolve(repoPath, filePath);

        let status: GitFileStatus = "?";
        let staged = false;

        if (indexStatus === "?" && workTreeStatus === "?") {
          status = "?";
        } else if (indexStatus === "A" || workTreeStatus === "A") {
          status = "A";
          staged = indexStatus === "A";
        } else if (indexStatus === "M" || workTreeStatus === "M") {
          status = "M";
          staged = indexStatus === "M";
        } else if (indexStatus === "D" || workTreeStatus === "D") {
          status = "D";
          staged = indexStatus === "D";
        } else if (indexStatus === "R" || workTreeStatus === "R") {
          status = "R";
        }

        const stats = numstatMap.get(filePath) || numstatMap.get(rawFilePath);

        files.push({
          path: absolutePath,
          status,
          staged,
          additions: stats?.additions ?? (status === "?" ? 0 : 0),
          deletions: stats?.deletions ?? 0,
        });
      }

      return {
        branch,
        clean: files.length === 0,
        files,
      };
    } catch {
      // Not a git repository or git command failed
      return {
        branch: "",
        clean: true,
        files: [],
      };
    }
  }

  /**
   * Get unified diff for the repository or a specific file.
   */
  async getDiff(repoPath: string, filePath?: string): Promise<string> {
    try {
      let cwd = repoPath;
      let targetPath = filePath;

      if (filePath && path.isAbsolute(filePath)) {
        try {
          const rootRes = await execAsync("git rev-parse --show-toplevel", {
            cwd: path.dirname(filePath),
          });
          if (rootRes.stdout.trim()) {
            cwd = rootRes.stdout.trim();
            targetPath = path.relative(cwd, filePath);
          }
        } catch {
          // If rev-parse fails, keep cwd as repoPath
        }
      }

      const target = targetPath ? ` -- "${targetPath}"` : "";

      // 1. Try staged + unstaged diff against HEAD
      const diffRes = await execAsync(`git diff HEAD${target}`, { cwd });
      if (diffRes.stdout.trim()) {
        return diffRes.stdout;
      }

      // 2. Try unstaged diff (e.g. in fresh repository without commits)
      const unstagedRes = await execAsync(`git diff${target}`, { cwd });
      if (unstagedRes.stdout.trim()) {
        return unstagedRes.stdout;
      }

      // 3. If file is untracked, produce unified diff against /dev/null
      if (targetPath) {
        try {
          const statusRes = await execAsync(`git status --porcelain -- "${targetPath}"`, { cwd });
          const statusLine = statusRes.stdout.trim();
          if (statusLine.startsWith("?") || statusLine.startsWith("A")) {
            const noIndexRes = await execAsync(
              `git diff --no-index /dev/null "${targetPath}"`,
              { cwd }
            ).catch((err: { stdout?: string }) => ({ stdout: err.stdout || "" }));
            if (noIndexRes.stdout.trim()) {
              return noIndexRes.stdout;
            }
          }
        } catch {}
      }

      return "";
    } catch {
      return "";
    }
  }

  /**
   * List all local branches, marking the checked-out one.
   *
   * A folder without Git is a valid project state. It returns an explicit
   * `isGitRepository: false` response instead of being indistinguishable from
   * an empty branch list or a failed request.
   */
  async listBranches(repoPath: string): Promise<GitBranchesResponse> {
    try {
      const repositoryCheck = await execAsync("git rev-parse --is-inside-work-tree", {
        cwd: repoPath,
      });
      if (repositoryCheck.stdout.trim() !== "true") {
        return { branches: [], isGitRepository: false };
      }

      const { stdout } = await execAsync('git branch --format="%(refname:short)"', {
        cwd: repoPath,
      });
      const branches = stdout
        .split("\n")
        .map((line) => line.trim())
        .filter(Boolean);
      let current = "";
      try {
        const currentRes = await execAsync("git rev-parse --abbrev-ref HEAD", {
          cwd: repoPath,
        });
        current = currentRes.stdout.trim();
      } catch {
        // Empty repository or detached HEAD: no current local branch.
      }
      return {
        branches: branches.map((name) => ({ name, current: name === current })),
        isGitRepository: true,
      };
    } catch (error) {
      if (isNotGitRepositoryError(error)) {
        return { branches: [], isGitRepository: false };
      }
      throw error;
    }
  }

  /**
   * Check out an existing local branch.
   */
  async checkoutBranch(repoPath: string, branch: string): Promise<void> {
    await execAsync(`git switch "${branch}"`, { cwd: repoPath });
  }
}

File: apps/server-go/internal/services/git_service.go
// Git operations via the git CLI. Port of
// apps/server/api/src/services/git.service.ts (status, diff, branches,
// checkout) with the same parsing rules.
package services

import (
	"os/exec"
	"path/filepath"
	"strconv"
	"strings"

	"github.com/Adelodunpeter25/console/apps/server-go/internal/types"
)

type GitService struct{}

func NewGitService() *GitService { return &GitService{} }

func runGit(dir string, args ...string) (string, error) {
	cmd := exec.Command("git", args...)
	cmd.Dir = dir
	out, err := cmd.Output()
	return string(out), err
}

// GetGitStatus returns branch + porcelain file entries with numstat deltas.
func (s *GitService) GetGitStatus(repoPath string) types.GitStatusSummary {
	branchOut, err := runGit(repoPath, "rev-parse", "--abbrev-ref", "HEAD")
	if err != nil {
		// Not a git repository or git failed — clean empty summary.
		return types.GitStatusSummary{Branch: "", Clean: true, Files: []types.GitFileEntry{}}
	}
	statusOut, _ := runGit(repoPath, "status", "--porcelain=v1", "-u")
	numstat := s.numstatMap(repoPath)

	branch := strings.TrimSpace(branchOut)
	if branch == "" {
		branch = "main"
	}
	files := make([]types.GitFileEntry, 0)
	for _, line := range strings.Split(statusOut, "\n") {
		if strings.TrimSpace(line) == "" || len(line) < 4 {
			continue
		}
		indexStatus, workTreeStatus := line[0], line[1]
		rawFilePath := strings.TrimSpace(line[3:])
		filePath := rawFilePath
		if idx := strings.Index(rawFilePath, "->"); idx >= 0 {
			filePath = strings.TrimSpace(rawFilePath[idx+2:])
		}
		absPath, _ := filepath.Abs(filepath.Join(repoPath, filePath))

		var status types.GitFileStatus = "?"
		staged := false
		switch {
		case indexStatus == '?' && workTreeStatus == '?':
			status = "?"
		case indexStatus == 'A' || workTreeStatus == 'A':
			status, staged = "A", indexStatus == 'A'
		case indexStatus == 'M' || workTreeStatus == 'M':
			status, staged = "M", indexStatus == 'M'
		case indexStatus == 'D' || workTreeStatus == 'D':
			status, staged = "D", indexStatus == 'D'
		case indexStatus == 'R' || workTreeStatus == 'R':
			status = "R"
		}
		stats, ok := numstat[filePath]
		if !ok {
			stats, ok = numstat[rawFilePath]
		}
		additions, deletions := int64(0), int64(0)
		if ok {
			additions, deletions = stats[0], stats[1]
		}
		files = append(files, types.GitFileEntry{
			Path: absPath, Status: status, Staged: staged,
			Additions: additions, Deletions: deletions,
		})
	}
	return types.GitStatusSummary{Branch: branch, Clean: len(files) == 0, Files: files}
}

func (s *GitService) numstatMap(repoPath string) map[string][2]int64 {
	m := make(map[string][2]int64)
	if out, err := runGit(repoPath, "diff", "--numstat"); err == nil {
		parseNumstat(out, m)
	}
	if out, err := runGit(repoPath, "diff", "--cached", "--numstat"); err == nil {
		parseNumstat(out, m)
	}
	return m
}

func parseNumstat(out string, m map[string][2]int64) {
	for _, line := range strings.Split(out, "\n") {
		fields := strings.Fields(strings.TrimSpace(line))
		if len(fields) < 3 {
			continue
		}
		var add, del int64
		// Binary files show "-" for both counts.
		if fields[0] != "-" {
			add, _ = strconv.ParseInt(fields[0], 10, 64)
		}
		if fields[1] != "-" {
			del, _ = strconv.ParseInt(fields[1], 10, 64)
		}
		m[fields[2]] = [2]int64{add, del}
	}
}

// GetDiff returns the unified diff for the repo or a single file.
func (s *GitService) GetDiff(repoPath string, filePath string) (string, error) {
	cwd := repoPath
	target := ""
	if filePath != "" {
		abs, _ := filepath.Abs(filePath)
		if repoPath == "" {
			if root, err := runGit(filepath.Dir(abs), "rev-parse", "--show-toplevel"); err == nil {
				cwd = strings.TrimSpace(root)
			} else {
				cwd = filepath.Dir(abs)
			}
		}
		target = " -- \"" + abs + "\""
	}
	if out, err := runGit(cwd, "diff", "HEAD"+strings.TrimPrefix(target, " --")); err == nil && out != "" {
		return out, nil
	}
	out, err := runGit(cwd, "diff")
	if err != nil {
		return "", err
	}
	if out == "" && filePath != "" {
		// Untracked file: synthesize a whole-file diff.
		if status, err := runGit(cwd, "status", "--porcelain", "--", filePath); err == nil && strings.TrimSpace(status) != "" {
			if noIndex, err := runGit(cwd, "diff", "--no-index", "/dev/null", filePath); err == nil {
				return noIndex, nil
			}
		}
	}
	return out, nil
}

// ListBranches lists local branches with the checked-out one marked.
func (s *GitService) ListBranches(repoPath string) types.GitBranchesResponse {
	if isRepo, err := runGit(repoPath, "rev-parse", "--is-inside-work-tree"); err != nil || strings.TrimSpace(isRepo) != "true" {
		return types.GitBranchesResponse{Branches: []types.GitBranchInfo{}, IsGitRepository: false}
	}
	out, _ := runGit(repoPath, "branch", "--format=%(refname:short)")
	currentOut, _ := runGit(repoPath, "rev-parse", "--abbrev-ref", "HEAD")
	current := strings.TrimSpace(currentOut)
	branches := make([]types.GitBranchInfo, 0)
	for _, name := range strings.Split(out, "\n") {
		name = strings.TrimSpace(name)
		if name == "" {
			continue
		}
		branches = append(branches, types.GitBranchInfo{Name: name, Current: name == current})
	}
	return types.GitBranchesResponse{Branches: branches, IsGitRepository: true}
}

func (s *GitService) CheckoutBranch(repoPath, branch string) error {
	_, err := runGit(repoPath, "switch", branch)
	return err
}
